hls-plugin-api/src/Ide/Plugin/Properties.hs

```diff
@@ -0,0 +1,341 @@
+{-# LANGUAGE AllowAmbiguousTypes  #-}
+{-# LANGUAGE ConstraintKinds      #-}
+{-# LANGUAGE DataKinds            #-}
+{-# LANGUAGE FlexibleContexts     #-}
+{-# LANGUAGE FlexibleInstances    #-}
+{-# LANGUAGE GADTs                #-}
+{-# LANGUAGE KindSignatures       #-}
+{-# LANGUAGE OverloadedStrings    #-}
+{-# LANGUAGE RecordWildCards      #-}
+{-# LANGUAGE ScopedTypeVariables  #-}
+{-# LANGUAGE TypeApplications     #-}
+{-# LANGUAGE TypeFamilies         #-}
+{-# LANGUAGE TypeOperators        #-}
+{-# LANGUAGE UndecidableInstances #-}
+-- See Note [Constraints]
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+
+module Ide.Plugin.Properties
+  ( PropertyType (..),
+    ToHsType,
+    MetaData (..),
+    PropertyKey,
+    SPropertyKey (..),
+    Properties,
+    HasProperty,
+    emptyProperties,
+    defineNumberProperty,
+    defineStringProperty,
+    defineBooleanProperty,
+    defineObjectProperty,
+    defineArrayProperty,
+    defineEnumProperty,
+    toDefaultJSON,
+    toVSCodeExtensionSchema,
+    usePropertyEither,
+    useProperty,
+    (&),
+  )
+where
+
+import qualified Data.Aeson       as A
+import qualified Data.Aeson.Types as A
+import           Data.Either      (fromRight)
+import           Data.Function    ((&))
+import           Data.Kind        (Constraint)
+import qualified Data.Map.Strict  as Map
+import           Data.Proxy       (Proxy (..))
+import qualified Data.Text        as T
+import           GHC.TypeLits
+import           Unsafe.Coerce    (unsafeCoerce)
+
+-- ---------------------------------------------------------------------
+
+data PropertyType
+  = TNumber
+  | TString
+  | TBoolean
+  | TObject
+  | TArray
+  | TEnum
+
+type family ToHsType (t :: PropertyType) where
+  ToHsType 'TNumber = Int
+  ToHsType 'TString = T.Text
+  ToHsType 'TBoolean = Bool
+  ToHsType 'TObject = A.Object
+  ToHsType 'TArray = A.Array
+  ToHsType 'TEnum = T.Text -- supports only text enum now
+
+-- ---------------------------------------------------------------------
+
+data MetaData (t :: PropertyType) where
+  MetaData ::
+    (IsNotTEnum t) =>
+    {defaultValue :: ToHsType t, description :: T.Text} ->
+    MetaData t
+  EnumMetaData ::
+    (IsTEnum t) =>
+    { defaultValue :: ToHsType t,
+      description :: T.Text,
+      enumValues :: [T.Text],
+      enumDescriptions :: [T.Text]
+    } ->
+    MetaData t
+
+data PropertyKey (s :: Symbol) (t :: PropertyType)
+
+data SPropertyKey k where
+  SNumber :: SPropertyKey (PropertyKey s 'TNumber)
```

Comment:
I don't understand why `s` is here --- presumably it's the name of the key, but I don't immediately see how it's specified or why it needs to be on this singleton.